### ISSUE TITLE
Map animation: optimize the TopoJSON map sequence

### ISSUE THREAD

**Issue by JohnTigue** [codeforseattle/seanetmap#71]
With any animation, the more "frames in the movie" the better. For SEANetMap's animated test results maps the more frames the better. Currently SEANetMap only does one frame per month and it is already hitting some performance issues. When we get to daily or even hourly frames we will be crushed.

Currently each time-frame in the movie is a TopoJSON file. These come out of Turf which does the grunt work of aggregating NDT test results by political boundary. Turf is outputing TopoJSON, which is a nice improvement over GeoJSON of earlier but for this discussion the difference is just an detail. 

The next step in the evolution would be to post-process the time-frame maps in order to remove the feature geometries and only use the feature's political boundary ID (US Census block group ID, City Council District ID, etc). Currently the geometries (which are the heavy part of Geo/TopoJSON files) are repeat in each time-frame map.  The idea is to only download the geometries once, since they do not change from time-frame to time-frame. **There is a huge optimization available here in both network and CPU load.**

So, there would be only one map with the (heavy) geometries and associated IDs. That map would have no performance report data in it -- it would just be a plain TopoJSON file with each feature having a well-known ID. All the time-frame maps would have the performance report data for that time period, but no geometries -- features would only be referenced by an ID. At this point the time-frame maps are really just JSON data-structs. They could even be CSVs but GeoJSON/TopoJSON is good enough and make the correspondence to the real Geo/TopoJSON files explicit and we do not need to come up with a CSV "schema." 

The client would read the time-frame report maps (or, for argument's sake, the CSVs but really don't because the Leaflet community will balk) and then change the color of the features on the cloropleth map to correspond to the report data (notice how Geo/TopoJSON maps are not being constantly re-parsed and rendered via the DOM, which has to be a good thing in terms of CPU load).

This has massive implications for how the map animation code works i.e. this is a costly issue to implement -- a total rebuild of the map animation machinery. I would not be surprised if someone has already worked this out in a reusable fashion. **If not then this is an extremely cool feature that the Leaflet community will be quite happy to use. I.e. high status for any motivated dev.**

For reference sake, here's the [GeoJSON spec](http://geojson.org/geojson-spec.html) for the terms used above.

**Comment by JohnTigue** [comment]
@nkinkade and @critzo I pulled this out of [the arch doc](https://github.com/codeforseattle/seanetmap/wiki/Architectural-overview) because I think it is a discrete issue that can be tracked. (And the arch doc is becoming a beast, too big for a wiki page.)

I am very much **not** implying that this should be worked on now. It's a big chunk of work for simply an animation optimization, where the animation currently works and many other things simply do not exist.

**Comment by JohnTigue** [comment]
Actually, keeping the aggregate report in CSV would do the job, be simple (adding a CSV parser is pretty darn simple with JS libraries already available, cleanly separate the TopoJSON aspect from the report aspect, would be more efficient over the wire.

Just saying it could go either way. (I do like the idea of featureless TopoJSON files but since the boundaries of the features do not change in our case, there really isn't any reason for TopoJSON wrapping of the report data. But it is a cool idea.)

So, in this version the resulting TopoJSON files would be generated (Turf is still needed to do the aggregation of points into features) but then there would be a post processing step which pulls the report data out of the `properties` and builds a CSV of the aggregate report.

**Comment by nkinkade** [comment]
One quick note on your comment "With any animation, the more "frames in the movie" the better. For SEANetMap's animated test results maps the more frames the better."  I'm not sure this is true, and in any case you will likely hit a problem:

1) The Internet is a highly variable thing.  Your performance will vary by minute, hour, day, month, year, etc., even with all other things being equal (e.g., computer, subscription level, ISP).  I think what is probably more Interesting, from a policy perspective, is not observing the hour/daily vagaries of connection speeds, but the more long term aggregation.  That is, viewing an animation of the, say, hourly aggregations is likely to hide more important trends... you can't see the forest through the trees... or you're seeing more noise than signal.

2) Even if you wanted to animate the map using hourly or daily aggregations, you would likely run into a test volume issue.  For any given polygon on a map, there are _months_ where the test count does not exceed a reasonable minimum to be statistically useful.  If you start getting down to the level of hours or days you are surely going to find that test volume is no where near high enough to provide meaningful data.  This, of course, can be surmounted by somehow dramatically increasing the test volume, but isn't the present state of things.

**Comment by nkinkade** [comment]
Regarding CSV files, as you point out, right now I'm doing all the aggregating on the server side.  But Turf.js will run in a browser too, so you could easily experiment with sending the browser the CSV file and trying to let Turf do all the processing on the client side.  Based on what I've seen on the server-side, I somewhat doubt this will prove to improve performance.  While a lot smaller than the processed file, you will still incur a time hit loading the CSV file across the network, it will then have to be converted to GeoJSON, which will incur a small time hit.  The GeoJSON will then have to be processed by Turf, which will likely incur _at least_ a two or three second hit, perhaps more if there happens to be a lot of data for that time period.  But it would be worth trying!

**Comment by JohnTigue** [comment]
I agree. There is no way I would want Turf processing this client side. Way too much compute and network hits.

The idea is to have all the Turf work still happen once per map frame generation step, not at "run time" i.e. while users are trying to consume the info. The Turf files would be post-processed server-side to jettison the weight of repeating all features in each time-frame report so that we are not dragging those across the network to be repeated re-rendered client side (with the exact same boundaries each time). 

This plan would still need [Turf server-side to do the aggregation](https://github.com/m-lab/bq2geojson/blob/master/index.js#L394):
``` 
var json = turf.aggregate(polygon, json, aggs);
```
Moving forward let us imaging the report data in CSV:
- it is easier for naming/discussing; no confusing "what is a featureless TopoJSON file?"
- this teases apart the roles
  - CSV for reporting 
  - GeoJSON/TopoJSON for providing feature shapes
- CSV is simply terser on the wire so it is a superior solution than reports embedded int TopoJSON 

So, server (not at run-time) generates only CSV reports, automatically on a regular basis. There is a run-of-the-mill TopoJSON (with feature geometries and feature IDs) is only generated once per, well, deploy. Looks like you already have feature IDs in there.

Client simply downloads one TopoJSON file once on init and reads&renders the TopoJSON once. Then if user wants to animate, the CSV report is used to change feature colors in a timer loop, just Leaflet and a timer no Turf.

**Comment by critzo** [comment]
We also should consider how postgres will/can do the aggregation. This is what David is planning. I'll share the repo where he's working as soon as I can.

**Comment by JohnTigue** [comment]
Regarding changing political boundaries:

This topic is way, way out into the future. Version 3.x or so?

Political boundaries do change, ever so often e.g. US Census gets redrawn every 10 years. A rock-solid implementation would need to address this, long-term.

One way to adress this is to have something (JSON, not Geo/TopoJSON) make the association between a CSV report and a TopoJSON. Internally the CSV cross-links to the TopoJSON via feature IDs. CSV has featureIdRef and TopoJSON has featureIDs.

So, an animation that handles changing political boundaries would have the following data structure
- At root is a sequence of CSV to TopoJSON associations. In JSON that would be an array.
- Each association has one static/stable/vanilla TopoJSON file (each feature having a featureID) for which its associated CSV report can use featureIdRefs
- Each CSV has a begin and end date range within which there are multiple reports for various intervals.

So for example if a deploy has data for Seattle 2014 to 2024, there would be a two element array
```
[
{"csv:"2014-2019.csv", "topojson":"us-census-block-groups-seattle-2010.json",
{"csv:"2020-2024.csv", "topojson":"us-census-block-groups-seattle-2020.json",
]
```

**Comment by JohnTigue** [comment]
@critzo:
Is David working on PostGIS doing the work Turf is doing in bq2geojson?
